File: Python/File_Converter.py
from __future__ import annotations

import argparse
import csv
import json
from pathlib import Path
from typing import Any


def read_text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def write_text(path: Path, content: str) -> None:
    path.write_text(content, encoding="utf-8")
    print(f"Wrote: {path}")


def csv_to_json(input_path: Path, output_path: Path) -> None:
    with input_path.open("r", encoding="utf-8", newline="") as file:
        reader = csv.DictReader(file)
        rows = list(reader)

    output_path.write_text(
        json.dumps(rows, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    print(f"Converted CSV -> JSON: {output_path}")


def json_to_csv(input_path: Path, output_path: Path) -> None:
    data: Any = json.loads(read_text(input_path))

    if not isinstance(data, list):
        raise ValueError("JSON must be a list of objects for json->csv conversion.")

    if len(data) == 0:
        output_path.write_text("", encoding="utf-8")
        print(f"Converted JSON -> CSV (empty): {output_path}")
        return

    if not all(isinstance(item, dict) for item in data):
        raise ValueError("All items in JSON array must be objects for json->csv.")

    fieldnames: list[str] = sorted(
        {key for item in data for key in item.keys()},
    )

    with output_path.open("w", encoding="utf-8", newline="") as file:
        writer = csv.DictWriter(file, fieldnames=fieldnames)
        writer.writeheader()
        for row in data:
            writer.writerow(row)

    print(f"Converted JSON -> CSV: {output_path}")


def txt_to_json(input_path: Path, output_path: Path) -> None:
    lines = [
        line.strip() for line in read_text(input_path).splitlines() if line.strip()
    ]

    output_path.write_text(
        json.dumps(lines, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    print(f"Converted TXT -> JSON: {output_path}")


def json_to_txt(input_path: Path, output_path: Path) -> None:
    data: Any = json.loads(read_text(input_path))

    if isinstance(data, list):
        content = "\n".join(str(item) for item in data)
    else:
        content = json.dumps(data, ensure_ascii=False, indent=2)

    write_text(output_path, content)
    print(f"Converted JSON -> TXT: {output_path}")


def convert(input_path: Path, output_path: Path) -> None:
    if not input_path.exists():
        raise FileNotFoundError(f"Input file not found: {input_path}")

    src = input_path.suffix.lower()
    dst = output_path.suffix.lower()

    if src == ".csv" and dst == ".json":
        csv_to_json(input_path, output_path)
        return

    if src == ".json" and dst == ".csv":
        json_to_csv(input_path, output_path)
        return

    if src == ".txt" and dst == ".json":
        txt_to_json(input_path, output_path)
        return

    if src == ".json" and dst == ".txt":
        json_to_txt(input_path, output_path)
        return

    raise ValueError(f"Unsupported conversion: {src} -> {dst}")


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Simple file converter (CSV/JSON/TXT).",
    )
    parser.add_argument(
        "--input",
        required=True,
        help="Input file path",
    )
    parser.add_argument(
        "--output",
        required=True,
        help="Output file path",
    )
    return parser


def main() -> None:
    parser = build_parser()
    args = parser.parse_args()

    try:
        convert(Path(args.input), Path(args.output))
    except Exception as error:
        print(f"Conversion failed: {error}")


if __name__ == "__main__":
    main()

File: Javascript/File_Converter.js
const fs = require('node:fs');
const path = require('node:path');

const parseArgs = () => {
  const args = process.argv.slice(2);
  const parsed = { input: null, output: null };

  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === '--input' && args[i + 1]) {
      parsed.input = args[i + 1];
      i += 1;
    } else if (args[i] === '--output' && args[i + 1]) {
      parsed.output = args[i + 1];
      i += 1;
    }
  }

  return parsed;
};

const readText = (filePath) => fs.readFileSync(filePath, 'utf8');

const writeText = (filePath, content) => {
  fs.writeFileSync(filePath, content, 'utf8');
  console.log(`Wrote: ${filePath}`);
};

const csvToJson = (inputPath, outputPath) => {
  const lines = readText(inputPath)
    .split(/\r?\n/)
    .filter((line) => line.trim() !== '');

  const headers = lines[0].split(',').map((header) => header.trim());
  const rows = lines.slice(1).map((line) => {
    const values = line.split(',').map((value) => value.trim());
    const row = {};
    headers.forEach((header, index) => {
      row[header] = values[index] ?? '';
    });
    return row;
  });

  writeText(outputPath, JSON.stringify(rows, null, 2));
  console.log(`Converted CSV -> JSON: ${outputPath}`);
};

const jsonToCsv = (inputPath, outputPath) => {
  const data = JSON.parse(readText(inputPath));
  if (!Array.isArray(data)) {
    throw new Error('JSON must be a list of objects for json->csv conversion.');
  }

  if (data.length === 0) {
    writeText(outputPath, '');
    console.log(`Converted JSON -> CSV (empty): ${outputPath}`);
    return;
  }

  const fieldNames = [...new Set(data.flatMap((item) => Object.keys(item)))].sort();
  const rows = [fieldNames.join(',')];

  for (const item of data) {
    const row = fieldNames.map((field) => String(item[field] ?? ''));
    rows.push(row.join(','));
  }

  writeText(outputPath, `${rows.join('\n')}\n`);
  console.log(`Converted JSON -> CSV: ${outputPath}`);
};

const txtToJson = (inputPath, outputPath) => {
  const lines = readText(inputPath)
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean);

  writeText(outputPath, JSON.stringify(lines, null, 2));
  console.log(`Converted TXT -> JSON: ${outputPath}`);
};

const jsonToTxt = (inputPath, outputPath) => {
  const data = JSON.parse(readText(inputPath));
  const content = Array.isArray(data)
    ? data.map((item) => String(item)).join('\n')
    : JSON.stringify(data, null, 2);

  writeText(outputPath, content);
  console.log(`Converted JSON -> TXT: ${outputPath}`);
};

const convert = (inputPath, outputPath) => {
  if (!fs.existsSync(inputPath)) {
    throw new Error(`Input file not found: ${inputPath}`);
  }

  const src = path.extname(inputPath).toLowerCase();
  const dst = path.extname(outputPath).toLowerCase();

  if (src === '.csv' && dst === '.json') {
    csvToJson(inputPath, outputPath);
    return;
  }

  if (src === '.json' && dst === '.csv') {
    jsonToCsv(inputPath, outputPath);
    return;
  }

  if (src === '.txt' && dst === '.json') {
    txtToJson(inputPath, outputPath);
    return;
  }

  if (src === '.json' && dst === '.txt') {
    jsonToTxt(inputPath, outputPath);
    return;
  }

  throw new Error(`Unsupported conversion: ${src} -> ${dst}`);
};

const main = () => {
  const args = parseArgs();

  if (!args.input || !args.output) {
    console.log('Usage: node File_Converter.js --input <file> --output <file>');
    return;
  }

  try {
    convert(args.input, args.output);
  } catch (error) {
    console.log(`Conversion failed: ${error.message}`);
  }
};

main();
